### PR TITLE
fix and test for issue 1758

### PR DESCRIPTION
**Pull request by gracekarina**
Fix and test for issue #1758.

Adds a validation message for parameters and headers when ref field in **internal references** are not being pointed at the right components tag.

### CODE REVIEW

File: modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/util/OpenAPIDeserializer.java [hunk 1]
```diff
@@ -1905,10 +1905,13 @@ public Parameter getParameter(ObjectNode obj, String location, ParseResult resul
...
 				if (mungedRef != null) {
-					parameter.set$ref(mungedRef);
-				} else {
-					parameter.set$ref(ref.textValue());
-				}
+                    parameter.set$ref(mungedRef);
+                }else {
+                    parameter.set$ref(ref.textValue());
+                }
+                if(ref.textValue().startsWith("#/components") && !(ref.textValue().startsWith("#/components/parameters")||ref.textValue().startsWith("#/components/headers"))) {
```
Comment (frantuma) [inline modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/util/OpenAPIDeserializer.java:1912]: Why is it accepting a ref to `#/components/headers` for parameters? (and the other way around below?)

**Review by frantuma — CHANGES_REQUESTED** [review]

File: modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/util/OpenAPIDeserializer.java [hunk 2]
```diff
@@ -1905,10 +1905,13 @@ public Parameter getParameter(ObjectNode obj, String location, ParseResult resul
...
+                    result.warning(location, "$ref is not pointing to #/components/parameters");
```
Comment (frantuma) [inline modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/util/OpenAPIDeserializer.java:1913]: possibly change to something like `$ref target #/components/headers/whatever is not of expected type 'header'`. @ponelat possibly has some even better wording

File: modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/util/OpenAPIDeserializer.java (hunk 1 above)
Comment (gracekarina) [reply modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/util/OpenAPIDeserializer.java:1912]: headers and parameters do share the same structure, and they can use with one or the other.
Docs: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.0.md#header-object
tests that use it: https://github.com/swagger-api/swagger-parser/blob/707adcb2fe3a2d31e8c915d9e4cacdb3b176f7c2/modules/swagger-parser-v3/src/test/java/io/swagger/v3/parser/test/OpenAPIV3ParserTest.java#L524

**Comment by gracekarina** [comment]
@ponelat @frantuma I also did the validation in deserializer of examples, and securitySchemes, but not 100% sure it applies here.
If you have a sample of a ref example being use, and securitySchemes using the ref `#/components/...` please let me know, I couldn't find any.
Thanks

**Comment by ponelat** [comment]
@gracekarina `securitySchemas` aren't referenced using `$ref`, instead they're referenced via the `security: []` fields in operations and globally.

Here is an example of using `$ref` to point to an example. Not that they're not for schemas, but for response/requestBodies, i.e., `examples`, see: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#exampleObject

```yaml
openapi: 3.0.3
info:
  title: $ref of an "Example".
  version: 1.0.0
paths:
  /foo:
    get:
      description: ok
      responses:
        '200':
          description: ok
          content:
            application/json:
              schema:
                type: object
              examples:
                one:
                  $ref: '#/components/examples/Ex'
              
components:
  examples:
    Ex:
      summary: The value one.
      value: 
        one: '1'
  ```

File: modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/util/OpenAPIDeserializer.java [hunk 3]
```diff
@@ -958,6 +958,9 @@ public PathItem getPathItem(ObjectNode obj, String location, ParseResult result)
 				} else {
 					pathItem.set$ref(ref.textValue());
 				}
+                if(ref.textValue().startsWith("#/components") && !(ref.textValue().startsWith("#/components/pathItems"))) {
+                    result.warning(location, "$ref target "+ref.textValue() +" is not of expected type");
```
Comment (frantuma) [inline modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/util/OpenAPIDeserializer.java:962]: I would detail what is the expected type, like `"$ref target "+ref.textValue() +" is not of expected type pathItem");`. Same applies to similar code below

File: modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/util/OpenAPIDeserializer.java [hunk 4]
```diff
@@ -1905,10 +1914,13 @@ public Parameter getParameter(ObjectNode obj, String location, ParseResult resul
...
 				if (mungedRef != null) {
-					parameter.set$ref(mungedRef);
-				} else {
-					parameter.set$ref(ref.textValue());
-				}
+                    parameter.set$ref(mungedRef);
+                }else {
+                    parameter.set$ref(ref.textValue());
+                }
+                if(ref.textValue().startsWith("#/components") && !(ref.textValue().startsWith("#/components/parameters")||ref.textValue().startsWith("#/components/headers"))) {
```
Comment (frantuma) [inline modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/util/OpenAPIDeserializer.java:1921]: cdoe style, would replace `meters")||ref.textVal` with `meters") || ref.textVal`

File: modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/util/OpenAPIDeserializer.java [hunk 5]
```diff
@@ -2115,6 +2127,10 @@ public Header getHeader(ObjectNode headerNode, String location, ParseResult resu
 				} else {
 					header.set$ref(ref.textValue());
 				}
+                if(ref.textValue().startsWith("#/components") && !(ref.textValue().startsWith("#/components/parameters")||ref.textValue().startsWith("#/components/headers"))) {
```
Comment (frantuma) [inline modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/util/OpenAPIDeserializer.java:2130]: cdoe style, would replace `meters")||ref.textVal` with `meters") || ref.textVal`

**Review by frantuma — COMMENTED** [review]
@gracekarina couple of minor comments, when addressed please merge

**Pull request merged**